File: legacy_reference/.agents/challenger/test_auth.py
import http.client
import urllib.parse
import uuid

def test_bypass():
    print("Testing auth bypass...")
    
    # Generate a random UUID for the fake cookie
    fake_uuid = str(uuid.uuid4())
    
    # 1. Normal request without cookie
    conn = http.client.HTTPConnection("127.0.0.1", 3015)
    conn.request("GET", "/api/generate-form")
    res1 = conn.getresponse()
    print(f"Normal request without cookie: {res1.status}")
    res1.read()
    
    # 2. Request with fake UUID cookie, normal host
    conn.request("GET", "/api/generate-form", headers={"Cookie": f"employee_auth={fake_uuid}"})
    res2 = conn.getresponse()
    print(f"Fake UUID, normal Host: {res2.status}")
    res2.read()
    
    # 3. Request with fake UUID cookie and injected Host header (attacker.com)
    headers3 = {
        "Cookie": f"employee_auth={fake_uuid}",
        "Host": "attacker.com"
    }
    conn.request("GET", "/api/generate-form", headers=headers3)
    res3 = conn.getresponse()
    print(f"Fake UUID, Host: attacker.com: {res3.status}")
    res3.read()
    
    # 4. Request with fake UUID cookie and injected Host header with port (attacker.com:80)
    # The middleware does: `http://127.0.0.1:${request.nextUrl.port || process.env.PORT || 3000}`
    # If we pass attacker.com:80, request.nextUrl.port is 80.
    headers4 = {
        "Cookie": f"employee_auth={fake_uuid}",
        "Host": "attacker.com:80"
    }
    conn.request("GET", "/api/generate-form", headers=headers4)
    res4 = conn.getresponse()
    print(f"Fake UUID, Host: attacker.com:80: {res4.status}")
    res4.read()

    # Wait, what if the SSRF was that we bypass the fetch entirely?
    # Or what if we point the SSRF to a port that returns 200 OK?
    # Let's see if NextJS parses attacker.com:80 as port 80.
    

if __name__ == "__main__":
    test_bypass()

File: legacy_reference/.agents/challenger/test_auth.js
const http = require('http');

async function runTests() {
  const targetHost = 'localhost';
  const targetPort = 3000; 
  const path = '/api/generate-form';
  const fakeUUID = '12345678-1234-1234-1234-123456789012'; // Random UUID

  console.log("=== Starting Auth Bypass Tests ===");

  // Test 1: Fake UUID cookie, no Host header injection
  console.log("\n[Test 1] Testing fake UUID cookie without Host header injection...");
  await new Promise((resolve) => {
    const req1 = http.request({
      host: targetHost,
      port: targetPort,
      path: path,
      method: 'GET',
      headers: {
        'Cookie': `employee_auth=${fakeUUID}`
      }
    }, (res) => {
      console.log(`Status Code: ${res.statusCode}`);
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        console.log(`Response: ${data}`);
        if (res.statusCode !== 401) {
          console.log("-> ❌ Test 1 Failed: Auth bypass successful with just fake cookie!");
        } else {
          console.log("-> ✅ Test 1 Passed: Request blocked.");
        }
        resolve();
      });
    });
    req1.on('error', (e) => {
        console.error(`Request error: ${e.message}`);
        resolve();
    });
    req1.end();
  });

  // Test 2: Fake UUID cookie WITH Host header injection
  const fakeServerPort = 3001;
  let ssrfCaught = false;
  const fakeServer = http.createServer((req, res) => {
    console.log(`\n[!] Fake server received request to: ${req.url}`);
    ssrfCaught = true;
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ user: { id: 'hacked' } }));
  });
  
  fakeServer.listen(fakeServerPort, async () => {
      console.log(`\nFake server listening on port ${fakeServerPort}`);

      console.log(`\n[Test 2] Testing fake UUID cookie WITH Host header injection (Host: 127.0.0.1:${fakeServerPort})...`);
      
      await new Promise((resolve) => {
        const req2 = http.request({
          host: targetHost,
          port: targetPort,
          path: path,
          method: 'GET',
          headers: {
            'Cookie': `employee_auth=${fakeUUID}`,
            'Host': `127.0.0.1:${fakeServerPort}` 
          }
        }, (res) => {
          console.log(`Status Code: ${res.statusCode}`);
          let data = '';
          res.on('data', chunk => data += chunk);
          res.on('end', () => {
            console.log(`Response: ${data}`);
            
            if (ssrfCaught) {
                console.log("-> ❌ Test 2 Failed: SSRF occurred! Request sent to injected Host.");
            } else if (res.statusCode !== 401) {
              console.log("-> ❌ Test 2 Failed: Auth bypass successful, but no SSRF caught.");
            } else {
              console.log("-> ✅ Test 2 Passed: Request blocked, no SSRF.");
            }
            resolve();
          });
        });
        req2.on('error', (e) => {
            console.error(`Request error: ${e.message}`);
            resolve();
        });
        req2.end();
      });
      
      console.log(`\n[Test 3] Testing with 'host' header injection (lowercase)...`);
      ssrfCaught = false; // reset
      await new Promise((resolve) => {
        const req3 = http.request({
          host: targetHost,
          port: targetPort,
          path: path,
          method: 'GET',
          headers: {
            'Cookie': `employee_auth=${fakeUUID}`,
            'host': `127.0.0.1:${fakeServerPort}` 
          }
        }, (res) => {
          console.log(`Status Code: ${res.statusCode}`);
          let data = '';
          res.on('data', chunk => data += chunk);
          res.on('end', () => {
            console.log(`Response: ${data}`);
            
            if (ssrfCaught) {
                console.log("-> ❌ Test 3 Failed: SSRF occurred! Request sent to injected host.");
            } else if (res.statusCode !== 401) {
              console.log("-> ❌ Test 3 Failed: Auth bypass successful, but no SSRF caught.");
            } else {
              console.log("-> ✅ Test 3 Passed: Request blocked, no SSRF.");
            }
            resolve();
          });
        });
        req3.on('error', (e) => {
            console.error(`Request error: ${e.message}`);
            resolve();
        });
        req3.end();
      });

      fakeServer.close();
      console.log("\n=== Tests Complete ===");
  });
}

runTests();
